pkg/apis/kubeflow.org/v1/common_types.go

```diff
@@ -0,0 +1,220 @@
+
+	// JobFailed means one or more sub-resources (e.g. services/pods) of this job
+	// reached phase failed with no restarting.
+	// The training has failed its execution.
+	JobFailed JobConditionType = "Failed"
+)
+
+// CleanPodPolicy describes how to deal with pods when the job is finished.
+type CleanPodPolicy string
+
+const (
+	CleanPodPolicyUndefined CleanPodPolicy = ""
+	CleanPodPolicyAll       CleanPodPolicy = "All"
+	CleanPodPolicyRunning   CleanPodPolicy = "Running"
+	CleanPodPolicyNone      CleanPodPolicy = "None"
+)
+
+// RestartPolicy describes how the replicas should be restarted.
+// Only one of the following restart policies may be specified.
+// If none of the following policies is specified, the default one
+// is RestartPolicyAlways.
+type RestartPolicy string
+
+const (
+	RestartPolicyAlways    RestartPolicy = "Always"
+	RestartPolicyOnFailure RestartPolicy = "OnFailure"
+	RestartPolicyNever     RestartPolicy = "Never"
+
+	// RestartPolicyExitCode policy means that user should add exit code by themselves,
+	// The job operator will check these exit codes to
+	// determine the behavior when an error occurs:
+	// - 1-127: permanent error, do not restart.
+	// - 128-255: retryable error, will restart the pod.
+	RestartPolicyExitCode RestartPolicy = "ExitCode"
+)
+
+// RunPolicy encapsulates various runtime policies of the distributed training
+// job, for example how to clean up resources and how long the job can stay
+// active.
+type RunPolicy struct {
+	// CleanPodPolicy defines the policy to kill pods after the job completes.
+	// Default to None.
+	CleanPodPolicy *CleanPodPolicy `json:"cleanPodPolicy,omitempty"`
+
+	// TTLSecondsAfterFinished is the TTL to clean up jobs.
+	// It may take extra ReconcilePeriod seconds for the cleanup, since
+	// reconcile gets called periodically.
+	// Default to infinite.
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+
+	// Specifies the duration in seconds relative to the startTime that the job may be active
+	// before the system tries to terminate it; value must be positive integer.
+	// +optional
+	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty"`
+
+	// Optional number of retries before marking this job failed.
+	// +optional
+	BackoffLimit *int32 `json:"backoffLimit,omitempty"`
+
+	// SchedulingPolicy defines the policy related to scheduling, e.g. gang-scheduling
+	// +optional
+	SchedulingPolicy *SchedulingPolicy `json:"schedulingPolicy,omitempty"`
+}
+
+// SchedulingPolicy encapsulates various scheduling policies of the distributed training
+// job, for example `minAvailable` for gang-scheduling.
+
+type SchedulingPolicy struct {
+	MinAvailable           *int32                                 `json:"minAvailable,omitempty"`
+	Queue                  string                                 `json:"queue,omitempty"`
+	MinResources           *map[v1.ResourceName]resource.Quantity `json:"minResources,omitempty"`
```

Comment:
Since this involves changing the API, we will do it in V2 version

Reply:
for more clalify.

If we use  `*v1.ResourceList`, the controller-gen goes to panic.

So we keep using `*map[v1.ResourceName]resourceQuantity` to avoid breaking changes although we shouldn't use type map as a pointer in the first place.

Reply:
Yes, we should use `corev1.ResourceList`, and then I confirm that if we use `core/ResourceList`, we don't face code-generator panic. But, changing `*corev1.ResourceList` (== `*map[v1.ResourceName]resourceQuantity`) to `corev1.ResourceList` has a breaking API change.

So, we decided to use `*map[corev1.ResourceName]resourceQuantity` in the v1 API upon talking with @johnugeorge offline.